util/conditions/patch.go

```diff
@@ -0,0 +1,140 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conditions
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+// PatchOperation defines patch operation types.
+type PatchOperation string
+
+const (
+	// AddConditionPatch defines an add condition patch operation.
+	AddConditionPatch PatchOperation = "Add"
+
+	// ChangeConditionPatch defines an change condition patch operation.
+	ChangeConditionPatch PatchOperation = "Change"
+
+	// RemoveConditionPatch defines a remove condition patch operation.
+	RemoveConditionPatch PatchOperation = "Remove"
+)
+
+// ConditionPatch define a patch that apply to a condition.
+type ConditionPatch struct {
+	Target *clusterv1.Condition
+	Base   *clusterv1.Condition
+	Op     PatchOperation
+}
+
+// GetPatch gets returns the list of ConditionPatch required to align source to target.
+func GetPatch(base clusterv1.Conditions, target clusterv1.Conditions) []ConditionPatch {
+	var patch []ConditionPatch
+
+	// Identify AddCondition and ModifyCondition changes-
+	for i := range target {
+		targetCondition := target[i]
+
+		currentCondition := get(base, targetCondition.Type)
+		if currentCondition == nil {
+			patch = append(patch, ConditionPatch{Op: AddConditionPatch, Target: &targetCondition})
+			continue
+		}
+
+		if !reflect.DeepEqual(&targetCondition, currentCondition) {
+			patch = append(patch, ConditionPatch{Op: ChangeConditionPatch, Target: &targetCondition, Base: currentCondition})
+		}
+	}
+
+	// Identify RemoveCondition changes-
+	for i := range base {
+		currentCondition := base[i]
+
+		targetCondition := get(target, currentCondition.Type)
+		if targetCondition == nil {
+			patch = append(patch, ConditionPatch{Op: RemoveConditionPatch, Base: &currentCondition})
+		}
+	}
+	return patch
+}
+
+// ApplyPatch executes a three way merge of a list of ConditionPatch.
+// When merge conflict are detected (source deviated from the original base recorded when creating the patch
+// in an incompatible way), an error is returned.
+func ApplyPatch(source clusterv1.Conditions, patch []ConditionPatch) (clusterv1.Conditions, error) {
+	merge := source.DeepCopy()
+
+	for i := range patch {
+		conditionPatch := patch[i]
+
+		switch conditionPatch.Op {
+		case AddConditionPatch:
+			// if the condition is already on source, check if source and target agree on the change;
+			// otherwise this is a conflict
+			if sourceCondition := get(source, conditionPatch.Target.Type); sourceCondition != nil {
+				// if source and target agree on the change, then it is a conflict
+				if !hasSameState(sourceCondition, conditionPatch.Target) {
+					return nil, errors.Errorf("error patching conditions: The condition %q on was modified by a different process and this caused a merge conflict", conditionPatch.Target.Type)
```

Comment:
I'm wondering if we should consider LastTransitionTime for the condition we are attempting to patch and allow updating the condition if our LastTransitionTime is newer than the LastTransitionTime from the source conditions?

Reply:
I struggled a little bit around the same problem and then I chose to not block for a small difference on LastTransitionTime while doing concurrent changes. Instead if there are changes in any other condition property, we are blocking. 
Also, I opted for preserving the first LastTransitionTime stored in etcd, so we avoid flickering of conditions value for a matter of ms.